xarray/core/rolling.py

```diff
@@ -236,31 +330,62 @@ def __init__(self, obj, min_periods=None, center=False, **windows):
         -------
         rolling : type of input argument
         """
-        from .dataset import Dataset
-        dim, window = next(iter(windows.items()))
-        # DataArrays not depending on the rolling-axis are kept aside from obj
-        self.fixed_ds = Dataset({key: da for key, da in obj.data_vars.items()
-                                 if dim not in da.dims},
-                                obj.coords)
-        # drop them from obj, then call parent's initializer.
-        obj = obj.drop(self.fixed_ds.data_vars.keys())
         super(DatasetRolling, self).__init__(obj,
                                              min_periods, center, **windows)
+        if self.dim not in self.obj.dims:
+            raise KeyError(self.dim)
+        # Keep each Rolling object as an OrderedDict
+        self.rollings = OrderedDict()
+        for key, da in self.obj.data_vars.items():
+            # keeps rollings only for the dataset depending on slf.dim
+            if self.dim in da.dims:
+                self.rollings[key] = DataArrayRolling(da, min_periods,
+                                                      center, **windows)
 
     def reduce(self, func, **kwargs):
-        if self.center:
-            fixed = self._center_result(self.fixed_ds)
-        else:
-            fixed = self.fixed_ds
-        # call parent's reduce
-        reduced = super(DatasetRolling, self).reduce(func, **kwargs)
-        # Need to restore dimension order
-        for key, var in reduced.data_vars.items():
-            reduced = reduced.assign(**{key:
-                                        var.transpose(*self.obj[key].dims)})
-        # merge the fixed DataArrays
-        return reduced.merge(fixed, join='inner')
+        """Reduce the items in this group by applying `func` along some
+        dimension(s).
+
+        Parameters
+        ----------
+        func : function
+            Function which can be called in the form
+            `func(x, **kwargs)` to return the result of collapsing an
+            np.ndarray over an the rolling dimension.
+        **kwargs : dict
+            Additional keyword arguments passed on to `func`.
 
+        Returns
+        -------
+        reduced : DataArray
+            Array with summarized data.
+        """
+        from .dataset import Dataset
+        reduced = OrderedDict()
+        for key, da in self.obj.data_vars.items():
+            if self.dim in da.dims:
+                reduced[key] = self.rollings[key].reduce(func, **kwargs)
+            else:
+                reduced[key] = self.obj[key]
+        return Dataset(reduced, coords=self.obj.coords)
+
+    @classmethod
+    def _reduce_method(cls, func):
+        """
+        Return a wrapped function for injecting numpy and bottoleneck methods.
+        see ops.inject_datasetrolling_methods
+        """
+        def wrapped_func(self, **kwargs):
```

Comment:
@shoyer 
Thank you for the review.
Could you give me more detail about the helper function?
I am wondering how I could distinguish the two types of reducing, 
`rolling.reduce(np.mean)` and `rolling.mean()`, the latter uses bottleneck if it is available.